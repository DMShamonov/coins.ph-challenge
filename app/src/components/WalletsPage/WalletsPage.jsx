// @flow

import * as React from 'react';

import {
  Page,
  PageHeader,
  PageContent,
  PageTitle,
} from 'components/Page';
import Flex from 'components/Flex';
import Total from 'components/Total';
import Button from 'components/Button';
import Wallets from 'components/Wallets';
import WalletCreateModal from 'components/WalletCreateModal';

import type { PropsType } from './types.js.flow';

export default function WalletsPage({
  wallets,
  onOpenCreateWalletModal,
}: PropsType): React.Node {
  return (
    <React.Fragment>
      <Page className="wallets-page">
        <PageHeader>
          <Flex>
            <PageTitle>Your wallets</PageTitle>
            <Total value={wallets.length} />
          </Flex>
          <Button onClick={onOpenCreateWalletModal}>Create wallet</Button>
        </PageHeader>
        <PageContent>
          {wallets.length ? <Wallets wallets={wallets} /> : 'You haven\'t wallets yet...'}
        </PageContent>
      </Page>
      <WalletCreateModal />
    </React.Fragment>
  );
}
