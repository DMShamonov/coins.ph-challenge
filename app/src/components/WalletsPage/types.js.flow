// @flow

import type { WalletsType } from 'types.js.flow';

export type PropsType = {|
  wallets: WalletsType,
  onOpenCreateWalletModal: () => void,
|};
