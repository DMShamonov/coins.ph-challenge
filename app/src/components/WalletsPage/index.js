import { compose } from 'redux';
import { connect } from 'react-redux';

import WALLETS_ACTION_TYPES from 'redux/actionTypes/wallets';
import MODAL_ACTION_TYPES from 'redux/actionTypes/modal';
import SESSION_ACTION_TYPES from 'redux/actionTypes/session';
import MODALS from 'constants/modals';
import preloader from 'decorators/preloader';
import api from 'controllers/api';

import WalletsPage from './WalletsPage';

function mapStateToProps({ data: { wallets } }) {
  return {
    wallets,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onOpenCreateWalletModal() {
      dispatch({ type: MODAL_ACTION_TYPES.ON_SHOW, payload: { modal: MODALS.CREATE_WALLET } });
    },
    dispatch,
  };
}

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  preloader(async ({ dispatch }) => {
    try {
      const { wallets: walletsList } = await api().wallets().list();
      const walletsData = await Promise.all(
        walletsList.map(({ wallet: { id } }) => api().wallets().get({ id })),
      );
      const wallets = walletsData.map(({ wallet }) => wallet);

      dispatch({ type: WALLETS_ACTION_TYPES.ON_SET, payload: { wallets } });
    } catch (e) {
      console.warn(e); // eslint-disable-line no-console

      if (e?.status === 401) {
        dispatch({ type: SESSION_ACTION_TYPES.LOGOUT });
      }
    }
  }),
)(WalletsPage);
