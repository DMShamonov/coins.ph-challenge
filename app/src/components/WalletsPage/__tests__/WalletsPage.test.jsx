import React from 'react';
import { shallow } from 'enzyme';

import WalletsPage from '../WalletsPage';

describe('`Wallet` component should render correctly', () => {
  test('with default props', () => {
    expect(shallow((
      <WalletsPage
        wallets={[{ id: '__ID__', label: '__LABEL__', balance: 100000000 }]}
        onOpenCreateWalletModal={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `wallets` prop is empty array', () => {
    expect(shallow((
      <WalletsPage
        wallets={[]}
        onOpenCreateWalletModal={jest.fn()}
      />
    ))).toMatchSnapshot();
  });
});
