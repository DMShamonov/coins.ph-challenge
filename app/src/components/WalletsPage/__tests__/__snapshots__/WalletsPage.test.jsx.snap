// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Wallet\` component should render correctly when \`wallets\` prop is empty array 1`] = `
<Fragment>
  <Page
    className="wallets-page"
  >
    <PageHeader>
      <Flex>
        <PageTitle>
          Your wallets
        </PageTitle>
        <Total
          value={0}
        />
      </Flex>
      <Button
        onClick={[MockFunction]}
      >
        Create wallet
      </Button>
    </PageHeader>
    <PageContent>
      You haven't wallets yet...
    </PageContent>
  </Page>
  <Connect(WalletCreateModal) />
</Fragment>
`;

exports[`\`Wallet\` component should render correctly with default props 1`] = `
<Fragment>
  <Page
    className="wallets-page"
  >
    <PageHeader>
      <Flex>
        <PageTitle>
          Your wallets
        </PageTitle>
        <Total
          value={1}
        />
      </Flex>
      <Button
        onClick={[MockFunction]}
      >
        Create wallet
      </Button>
    </PageHeader>
    <PageContent>
      <Wallets
        wallets={
          Array [
            Object {
              "balance": 100000000,
              "id": "__ID__",
              "label": "__LABEL__",
            },
          ]
        }
      />
    </PageContent>
  </Page>
  <Connect(WalletCreateModal) />
</Fragment>
`;
